### DISCUSSION TITLE
will be vlang ndk bind or whole build?

### DISCUSSION THREAD

**Discussion by MatejMagat305** [vlang/v#16809] [Questions and Answers] [answered]
It is far future, but I m curious so I ask, 
will be vlang ndk "bind" with "javaNativeActivity" like now and similar https://github.com/golang/mobile "build"
or 
whole build like  https://github.com/xlab/android-go without java on start (not without jni for some funkcion)?

Do you think: Vlang will be able run on mobile (ios too) as well as destop? Becouse mobile is more and more offen .... 

And last question, I have create jni interface for bluetooth RC connection for go can I sugest to vlang too ?

**Comment by JalonSolov** [comment]
V already builds fine for Android.  It builds for ios, but have not been able to successfully sign it the way Apple wants (since that doesn't seem to work unless you do it on a Mac with their XCode IDE).

See https://github.com/vlang/vab for V's Android support.

**Reply by Wajinn** [reply]
> V already builds fine for Android. It builds for ios, but have not been able to successfully sign it the way Apple wants (since that doesn't seem to work unless you do it on a Mac with their XCode IDE).
> 

If V can build it, and just needs to sign it, then that could possibly be done through VMware, VirtualBox, or renting a Mac over the cloud.  It could be at least good as an experiment that documents the process and then shows other V users the likely steps they would need to go through.

**Comment by larpon** [comment, marked as answer]
I do not know how any of the Go projects work with Android, so I can't answer to that specifically.

I do know that V runs fine on Android - and with [vab](https://github.com/vlang/vab) it works via [`NativeActivity`](https://github.com/vlang/vab/blob/a8352208b12694156c8b08d5b528d56861831e5c/platforms/android/src/io/v/android/VActivity.java#L3), vab allows you to use the [jni](https://github.com/Larpon/jni) but also requires Java, SDK and NDK to be installed on the host computer. V itself can be made to run on Android *without* Java - vab was actually based on [this approach](https://github.com/cnlohr/rawdrawandroid) in the early versions - but since the approach has many downsides in real world applications, vab changed to use a Java based approach.

In regards to iOS; yes, V should be able to run on iOS but the last attempts we made ended in a black screen and no further attempts has been made since then AFAIK. Apple doesn't provide very friendly nor very clear  documentation on such matters so progress is inherently slower with such a closed platform as iOS.

As for your last question, I'm not 100% what you mean, but you should be able to port any JNI code fairly easy to V, you can use my personal wrapper [jni](https://github.com/Larpon/jni) or make a wrapper yourself, if you so wish (I don't know of any other JNI wrappers for V than my own)

I hope that answers your questions

**Reply by Wajinn** [reply]
> In regards to iOS; yes, V should be able to run on iOS but the last attempts we made ended in a black screen and no further attempts has been made since then AFAIK. Apple doesn't provide very friendly nor very clear documentation on such matters so progress is inherently slower with such a closed platform as iOS.

Definitely would like to see team V master iOS, otherwise it could be seen as a weak point.  Maybe team V might want to track some Go solutions, as it appears they have some packages for iOS (can't say how well they work).  At least it might give some additional ideas to how to go about it or get around some problems.

On a side note, it seems another strong barrier is often developers who use Windows, Linux, or Android do not have Macs and iPhones to do testing on.  Probably the only cost effective way around that (though somewhat controversial) is running macOS (and likely Xcode) on VMware or VirtualBox.  Then there is renting a mac over the cloud, though not as cost effective for various people.

**Reply by larpon** [reply]
IIRC, V was able to print output to the console so it did run - the graphics backend just didn't show anything at the time. This was many many many months ago so I can't remember the specifics.

Yes devs could probably use VMs just like some do with Windows. It's just that; Apple is just not FOSS developer friendly, so I can see why open source devs might not be very motivated to do anything on the platform. One thing is getting things to build and develop locally... releasing to the app store and going through the whole app signing and notarization process is a whole chapter of pain in itself... 

For me personally it's also about an issue where Apple has locked me out of their development program for reasons unknown to me. I own both a mac and iPhone hardware - but their broken DRM has prevented me from developing anything for years. It just doesn't work on my machine for some reason 🤷🏻‍♂️ I even had to give up maintenance on the few apps I had in their app store, so they've now excluded them from being listed since I couldn't update them to meet their latest requirements (the requirements are updated quite frequently, btw).

I'm sure we'll get there eventually, though 🙃

**Reply by MatejMagat305** [reply]
Thank for antwere, that is good reason to use java ..., (I do not like not "alternative" to java on android [it is many, but at startup is java], but it is only my opinion [which is controversal]), I was only curious ..., by the way (btw) in my opinion, library in golang fyne.io have a good solution to solve java..., compile opcode is "embed" to building toolchain ..., but I m interasting to Vlang how will continious in future...

**Reply by Wajinn** [reply]
> the graphics backend just didn't show anything at the time...

That part seems interesting, as it would appear to relate to gg and the create_image functions.  Definitely would be interested if team V finds there still is an issue or its been resolved.

> Apple...

They definitely don't make it easy and were keen on maintaining their walled garden.  But, reading the news, it appears Apple will be forced to allow alternative app stores, as the result of new EU regulation coming soon.  This should make the situation more appealing to various developers.

**Reply by larpon** [reply]
It was not an image related issue but rather getting sokol to render *anything* as I remember it

**Reply by Wajinn** [reply]
> It was not an image related issue but rather getting sokol to render _anything_ as I remember it

Well, it doesn't appear that it's a matter that sokol can't do it, because there are a number of sokol demo apps on iOS.  Hopefully, when you guys get a chance, it can be looked at again and resolved.  

Many would argue that iOS is as important as android.  And with pressure on Apple to change up on their walled garden and allow alternative app stores, it should become more interesting.

**Comment by MatejMagat305** [comment]
Thank for everything, I will watch vlang and if I will be able I will try to improve something..., if I can I will have 2(3) other question..., 
1. if I do something small would I create PR or isues for example if I create derivetion of vsl.vcl for dlopen (it will realy small I predict) would I sugest PR or not?
2. is flag '-native' for vab android and how is it about 'hello_world' binary size ?

**Reply by JalonSolov** [reply]
1. An issue to suggest, a PR to implement.
2. `-native` only works for normal desktop binaries, and is still :construction: for now.  Perhaps some day it could be enhanced to work for Android as well, if someone is willing to put in the work to make it happen.

**Reply by MatejMagat305** [reply]
thank
